From: Mario Limonciello <supermario@mlimonciello.(none)>
Date: Tue, 27 May 2008 17:15:56 +0000 (-0500)
Subject: UBUNTU: SAUCE: Work around ACPI corruption upon suspend on some Dell machines.
X-Git-Tag: Ubuntu-2.6.24-19.33~16
X-Git-Url: http://kernel.ubuntu.com/git?p=ubuntu%2Fubuntu-hardy.git;a=commitdiff_plain;h=8a8edfd73da81c311e15ffba945a7b8eeb7f992f

UBUNTU: SAUCE: Work around ACPI corruption upon suspend on some Dell machines.

OriginalAuthor: Dennis Noordsij <dennis.noordsij@helsinki.fi>
OriginalLocation: http://bugzilla.kernel.org/attachment.cgi?id=16280&action=view
Bug: #183033
Ignore: no

Allocate ACPI descriptors instead of mapping ACPI memory.
This will resolve issues with CPU frequency scaling going out of wack upon
suspend and resume.

Signed-off-by: Mario Limonciello <Mario_Limonciello@.Dell.com>
Signed-off-by: Tim Gardner <tim.gardner@canonical.com>
---

Index: 2008-devel/drivers/acpi/executer/exconfig.c
===================================================================
--- 2008-devel.orig/drivers/acpi/executer/exconfig.c
+++ 2008-devel/drivers/acpi/executer/exconfig.c
@@ -268,6 +268,8 @@ acpi_ex_load_op(union acpi_operand_objec
 	struct acpi_table_desc table_desc;
 	acpi_native_uint table_index;
 	acpi_status status;
+	u32 length;
+	void *maddr;
 
 	ACPI_FUNCTION_TRACE(ex_load_op);
 
@@ -299,9 +301,24 @@ acpi_ex_load_op(union acpi_operand_objec
 			}
 		}
 
+ 		length = obj_desc->region.length;
+ 		table_desc.pointer = ACPI_ALLOCATE(length);
+ 		if (!table_desc.pointer) {
+ 			return_ACPI_STATUS(AE_NO_MEMORY);
+ 		}
+
+ 		maddr = acpi_os_map_memory(obj_desc->region.address, length);
+ 		if (!maddr) {
+ 			ACPI_FREE(table_desc.pointer);
+ 			return_ACPI_STATUS(AE_NO_MEMORY);
+ 		}
+ 		ACPI_MEMCPY(table_desc.pointer, maddr, length);
+ 		acpi_os_unmap_memory(maddr, length);
+
+ 		/* Keep the address for the pretty table info print */
 		table_desc.address = obj_desc->region.address;
 		table_desc.length = obj_desc->region.length;
-		table_desc.flags = ACPI_TABLE_ORIGIN_MAPPED;
+		table_desc.flags = ACPI_TABLE_ORIGIN_ALLOCATED;
 		break;
 
 	case ACPI_TYPE_BUFFER:	/* Buffer or resolved region_field */
